File: benchmarks/multilang-wasm/game-canvas-arcade/arcade_kernel.ts
// arcade_kernel.ts — AssemblyScript multilang compute core for
// game.canvas-arcade.v1. Same ABI as arcade_kernel.c: the adapter writes the
// frozen 14,424-byte arcade fixture at FIXTURE_OFFSET (65 536), passes the
// byte length, and this kernel replays the 3,600-frame arcade engine
// (bit-identical to run_arcade() in benchmarks/v2/game-family/game-family.c
// and arcade() in engine.js), writing counters + digests to RES_OFFSET
// (131 072). Raw linear-memory access only (no heap allocation, no runtime
// imports) — mirrors gc_document_kernel.ts.

const FIXTURE_OFFSET: usize = 65536;
const RES_OFFSET: usize = 131072;

function fixtureAt(off: u32): u8 {
  return load<u8>(FIXTURE_OFFSET + (<usize> off));
}
function read16(at: u32): u32 {
  return (<u32> fixtureAt(at)) | ((<u32> fixtureAt(at + 1)) << 8);
}
function read32(at: u32): u32 {
  return read16(at) | (read16(at + 2) << 16);
}
function mix(h: u32, v: u32): u32 {
  return ((h ^ v) * 0x01000193) >>> 0;
}

export function arcade_trace(fixture_len: u32): i32 {
  if (fixture_len != 14424) return 1;
  if (read32(0) != 3600) return 2;

  let state: u32 = 0x54a1c9e7;
  let draw: u32 = 0x9e3779b9;
  let audio: u32 = 0x243f6a88;
  let x: i32 = 640;
  let y: i32 = 600;
  let score: u32 = 0;
  let lives: u32 = 3;
  let entityUpdates: u32 = 0;
  let collisionTests: u32 = 0;
  let drawCommands: u32 = 0;
  let audioEvents: u32 = 0;

  for (let frame: u32 = 0; frame < 3600; frame++) {
    const control = read32(24 + frame * 4);
    const dx: i32 = ((control & 1) ? -7 : 0) + ((control & 2) ? 7 : 0);
    x = (x + dx + 1280) % 1280;
    const dy: i32 = ((control & 4) ? -5 : 0) + ((control & 8) ? 5 : 0);
    y += dy;
    if (y < 0) y = 0;
    if (y > 719) y = 719;
    const active: u32 = 32 + ((control >>> 8) & 31);
    draw = mix(mix(mix(draw, 0), frame), 0x050002d0);
    drawCommands++;
    for (let entity: u32 = 0; entity < active; entity++) {
      state = mix(state, frame * 131 + entity * 17 + control);
      entityUpdates++;
      state = mix(state, (<u32> (x + y)) + entity);
      collisionTests++;
      draw = mix(mix(mix(mix(draw, 2), frame), entity), state);
      drawCommands++;
      if ((state & 2047) == 0) {
        score += 10;
        audio = mix(mix(mix(audio, 1), frame), entity);
        audioEvents++;
      }
    }
    draw = mix(mix(mix(mix(draw, 1), frame), <u32> x), <u32> y);
    draw = mix(mix(mix(draw, 3), score), lives);
    drawCommands += 2;
    if ((control & 0xff00) == 0xff00 && lives > 0) {
      lives--;
      audio = mix(mix(mix(audio, 2), frame), lives);
      audioEvents++;
    }
    state = mix(state, (<u32> x) ^ ((<u32> y) << 11) ^ score ^ lives);
  }

  const semantic = mix(mix(state, draw), audio);
  store<u32>(RES_OFFSET, semantic);
  store<u32>(RES_OFFSET + 4, state);
  store<u32>(RES_OFFSET + 8, draw);
  store<u32>(RES_OFFSET + 12, audio);
  store<u32>(RES_OFFSET + 16, entityUpdates);
  store<u32>(RES_OFFSET + 20, collisionTests);
  store<u32>(RES_OFFSET + 24, drawCommands);
  store<u32>(RES_OFFSET + 28, audioEvents);
  return 0;
}

File: benchmarks/multilang-wasm/game-canvas-arcade/arcade_kernel.c
// arcade_kernel.c — multilang compute core for game.canvas-arcade.v1.
//
// ABI (mirrors the gc_document_kernel multilang template): the adapter writes
// the frozen 14,424-byte arcade fixture (game-v2-controlled-family/
// game-canvas-arcade-v1.bin — 6×u32 header + 3,600 u32 input events, generated
// by benchmarks/v2/game-family/fixtures.js with seed 0x6d2b79f5) into linear
// memory at FIXTURE_OFFSET, passes the byte length, and this kernel replays
// the 3,600-frame arcade engine (bit-for-bit identical to run_arcade() in
// benchmarks/v2/game-family/game-family.c and arcade() in engine.js). We
// accept the fixture bytes via memory rather than regenerating them, because
// the fixture is a 14 KiB seed-derived xorshift stream and the game-family
// authoritative C already reads it the same way — reusing that oracle keeps
// this kernel byte-identical with runGameJavaScript() and runGameWasm().
//
// Results (fixed offset RES_OFFSET, u32 little-endian):
//   [0] semanticDigest           (0x585a29e5)
//   [1] finalStateDigest         (0x87695460)
//   [2] drawCommandStreamDigest  (0xf3a03070)
//   [3] audioEventStreamDigest   (0x8b4cb497)
//   [4] entityUpdates            (169501)
//   [5] collisionTests           (169501)
//   [6] drawCommands             (180301)
//   [7] audioEvents              (91)
// Exports: i32 arcade_trace(u32 fixture_len) -> 0 on success, non-zero on
// fixture-length mismatch (bad input).

#define FIXTURE_OFFSET 65536
#define RES_OFFSET 131072

typedef unsigned int u32;
typedef int i32;
typedef unsigned char u8;

static u8 fixture_at(u32 off) { return *((u8 *)(FIXTURE_OFFSET) + off); }
static u32 read16(u32 at) {
  return (u32)fixture_at(at) | ((u32)fixture_at(at + 1) << 8);
}
static u32 read32(u32 at) { return read16(at) | (read16(at + 2) << 16); }
static u32 mix(u32 h, u32 v) { return (h ^ v) * 16777619u; }

__attribute__((export_name("arcade_trace")))
int arcade_trace(u32 fixture_len) {
  if (fixture_len != 14424u) return 1;
  if (read32(0) != 3600u) return 2;

  u32 state = 0x54a1c9e7u, draw = 0x9e3779b9u, audio = 0x243f6a88u;
  i32 x = 640, y = 600;
  u32 score = 0, lives = 3;
  u32 entity_updates = 0, collision_tests = 0, draw_commands = 0, audio_events = 0;

  for (u32 frame = 0; frame < 3600u; frame++) {
    const u32 control = read32(24u + frame * 4u);
    x = (x + ((control & 1u) ? -7 : 0) + ((control & 2u) ? 7 : 0) + 1280) % 1280;
    y += ((control & 4u) ? -5 : 0) + ((control & 8u) ? 5 : 0);
    if (y < 0) y = 0;
    if (y > 719) y = 719;
    const u32 active = 32u + ((control >> 8) & 31u);
    draw = mix(mix(mix(draw, 0u), frame), 0x050002d0u);
    draw_commands++;
    for (u32 entity = 0; entity < active; entity++) {
      state = mix(state, frame * 131u + entity * 17u + control);
      entity_updates++;
      state = mix(state, (u32)(x + y) + entity);
      collision_tests++;
      draw = mix(mix(mix(mix(draw, 2u), frame), entity), state);
      draw_commands++;
      if ((state & 2047u) == 0u) {
        score += 10u;
        audio = mix(mix(mix(audio, 1u), frame), entity);
        audio_events++;
      }
    }
    draw = mix(mix(mix(mix(draw, 1u), frame), (u32)x), (u32)y);
    draw = mix(mix(mix(draw, 3u), score), lives);
    draw_commands += 2u;
    if ((control & 0xff00u) == 0xff00u && lives > 0u) {
      lives--;
      audio = mix(mix(mix(audio, 2u), frame), lives);
      audio_events++;
    }
    state = mix(state, (u32)x ^ ((u32)y << 11) ^ score ^ lives);
  }

  const u32 semantic = mix(mix(state, draw), audio);
  u32 *results = (u32 *)RES_OFFSET;
  results[0] = semantic;
  results[1] = state;
  results[2] = draw;
  results[3] = audio;
  results[4] = entity_updates;
  results[5] = collision_tests;
  results[6] = draw_commands;
  results[7] = audio_events;
  return 0;
}
